### ISSUE TITLE
ReportingParticipant reentrancy bug for minting REP.

### ISSUE THREAD

**Issue by MicahZoltu** [AugurProject/augur#5124]
TL;DR: `mint` calls into user code and thus is a reentrancy vector.  `BaseReportingParticipante.fork` calls `REP.mint` prior to updating state that would prevent reentrant calls.

1. User calls `BaseReportingParticipant.fork`
1. --> Check is made against to verify the participant's market is the forking market for the market's universe.
1. `ReputationToken.migrateOutByPayout` is called to migrate the `oldREP` balance of the contract to `newREP`.
1. `ReputationToken.mintForReportingParticipant` is called to mint `newREP` for the participant.
1. --> Callback to user is made at this point (see #5123 for details).
1. The user can transfer 1 atto `oldREP` to the participant contract and then call `BaseReportingParticipant.fork` again.  Since the contract has REP, `migrateOutByPayout` will not fail and `mintForReportingParticipant` will get called again.
1. During unwind, `fork` will set `market = 0`, which will trigger protection, but it is too late.

Recommended fix: Move the `market = IMarket(0)` assignment to right after the `require(market == ...)` check.
